File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = {}
        left=0
        right=0
        res = 0

        while right < len(s):
            c = s[right]
            window[c] = window.get(c, 0) + 1
            right += 1

            while window[c] > 1:
                d = s[left]
                window[d] -= 1
                left += 1

            res = max(res, right - left)

        return res

File: Data Structures & Algorithms/is-palindrome/submission-5.java
class Solution {
    public boolean isPalindrome(String s) {
        if(s == null || s.length() == 0) return true;

        int left = 0;
        int right = s.length() - 1;

        while(left < right){
            while (left < right) {
                if (!Character.isLetterOrDigit(s.charAt(left))) {
                    left++;
                    } else {
                        break; // 找到有效字符后跳出循环
                        }
                        }
            while (left < right) {
                if (!Character.isLetterOrDigit(s.charAt(right))) {
                    right--;
                    } else {
                        break; // 找到有效字符后跳出循环
                        }
                        }


            char leftChar = s.charAt(left);
            char rightChar = s.charAt(right);
            if (Character.toLowerCase(leftChar) != Character.toLowerCase(rightChar)) {
                return false;
                }

            left++;
            right--;
        }

        return true;
    }
}//n,1
//125